crates/cranelift/src/compiler/component.rs

```diff
@@ -262,7 +328,337 @@ impl<'a> TrampolineCompiler<'a> {
+            &WasmFuncType::new(
+                Box::new([]),
+                wasm_func_ty.returns().iter().copied().collect(),
+            ),
+            &mut self.builder,
+            &[],
+        );
+        let values_vec_len = self.builder.ins().iconst(pointer_type, i64::from(len));
+
+        callee_args.push(values_vec_ptr);
+        callee_args.push(values_vec_len);
+
+        let call = self.call_libcall(vmctx, host::sync_exit, &callee_args);
+
+        let succeeded = self.builder.func.dfg.inst_results(call)[0];
+        self.raise_if_host_trapped(succeeded);
+        // After the host function has returned the results are loaded from
+        // `values_vec_ptr` and then returned.
+        let results = self.compiler.load_values_from_array(
+            wasm_func_ty.returns(),
+            &mut self.builder,
+            values_vec_ptr,
+            values_vec_len,
+        );
+        self.builder.ins().return_(&results);
+    }
+
+    fn translate_async_enter_or_exit(
+        &mut self,
+        get_libcall: fn(
+            &dyn TargetIsa,
+            &mut ir::Function,
+        ) -> (ir::SigRef, ComponentBuiltinFunctionIndex),
+        callback_and_post_return: Option<(
+            Option<RuntimeCallbackIndex>,
+            Option<RuntimePostReturnIndex>,
+        )>,
+        result: ir::types::Type,
+    ) {
+        match self.abi {
+            Abi::Wasm => {}
+
+            Abi::Array => {
+                // This code can only be called from (FACT-generated) Wasm, so
+                // we don't need to support the array ABI.
+                self.builder.ins().trap(TRAP_INTERNAL_ASSERT);
+                return;
+            }
+        }
+
+        let args = self.builder.func.dfg.block_params(self.block0).to_vec();
+        let vmctx = args[0];
+
+        let mut callee_args = vec![vmctx];
+
+        if let Some((callback, post_return)) = callback_and_post_return {
+            let pointer_type = self.isa.pointer_type();
+
+            // callback: *mut VMFuncRef
+            if let Some(callback) = callback {
+                callee_args.push(self.builder.ins().load(
+                    pointer_type,
+                    MemFlags::trusted(),
+                    vmctx,
+                    i32::try_from(self.offsets.runtime_callback(callback)).unwrap(),
+                ));
+            } else {
+                callee_args.push(self.builder.ins().iconst(pointer_type, 0));
+            }
+
+            // post_return: *mut VMFuncRef
+            if let Some(post_return) = post_return {
+                callee_args.push(self.builder.ins().load(
+                    pointer_type,
+                    MemFlags::trusted(),
+                    vmctx,
+                    i32::try_from(self.offsets.runtime_post_return(post_return)).unwrap(),
+                ));
+            } else {
+                callee_args.push(self.builder.ins().iconst(pointer_type, 0));
+            }
```

Comment:
Similar to down below handling these canonical ABI arguments I think would be good to have as helper functions